{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e8e3de6-8e29-4598-9917-8aa8450bede3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import pandas as pd\n",
    "from components import db_functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c7b86807",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join('data','db','db build files','uniprotkb_AND_reviewed_true_2023_09_04.tsv'),sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2f397e8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accession,reviewed,gene_primary,id,gene_names,organism_name,length,cc_alternative_products,ft_var_seq,sequence\n"
     ]
    }
   ],
   "source": [
    "map = {'Entry': 'accession',\n",
    "            'Post-translational modification': 'cc_ptm',\n",
    "            'Comment Count': 'comment_count',\n",
    "            'Organism': 'organism_name',\n",
    "            'Modified residue': 'ft_mod_res',\n",
    "            'Gene names': 'gene_names',\n",
    "            'Gene Names': 'gene_names',\n",
    "            'Gene names (primary)': 'gene_primary',\n",
    "            'Gene Names (primary)': 'gene_primary',\n",
    "            'Gene ontology (cellular component)': 'go_c',\n",
    "            'Gene ontology (molecular function)': 'go_f',\n",
    "            'Gene ontology (biological process)': 'go_p',\n",
    "            'Entry name': 'id',\n",
    "            'Entry Name': 'id',\n",
    "            'Keywords': 'keyword',\n",
    "            'Length': 'length',\n",
    "            'PubMed ID': 'lit_pubmed_id',\n",
    "            'Mass': 'mass',\n",
    "            'Protein names': 'protein_name',\n",
    "            'Reviewed': 'reviewed',\n",
    "            'Sequence': 'sequence',\n",
    "            'Entry version': 'version',\n",
    "            'Reactome': 'xref_reactome',\n",
    "            'Cofactor': 'cc_cofactor',\n",
    "            'Developmental stage': 'cc_developmental_stage',\n",
    "            'Involvement in disease': 'cc_disease',\n",
    "            'Disruption phenotype': 'cc_disruption_phenotype',\n",
    "            'Domain[CC]': 'cc_domain',\n",
    "            'Function [CC]': 'cc_function',\n",
    "            'Induction': 'cc_induction',\n",
    "            'Interacts with': 'cc_interaction',\n",
    "            'Miscellaneous [CC]': 'cc_miscellaneous',\n",
    "            'Sequence caution': 'cc_sequence_caution',\n",
    "            'Subcellular location[CC]': 'cc_subcellular_location',\n",
    "            'Date of last sequence modification': 'date_sequence_modified',\n",
    "            'Organism ID': 'organism_id',\n",
    "            'PhosphoSitePlus': 'xref_phosphositeplus',\n",
    "            'Sequence version': 'sequence_version',\n",
    "            'RefSeq': 'xref_refseq',\n",
    "            'STRING': 'xref_string',\n",
    "            'Ensembl': 'xref_ensembl',\n",
    "            'IntAct': 'xref_intact',\n",
    "            'InterPro': 'xref_interpro',\n",
    "            'PANTHER': 'xref_panther',\n",
    "            'SUPFAM': 'xref_supfam',\n",
    "            'Pfam': 'xref_pfam',\n",
    "            'PRO': 'xref_pro',\n",
    "            'PROSITE': 'xref_prosite',\n",
    "            'Alternative sequence': 'ft_var_seq',\n",
    "            'Alternative products (isoforms)': 'cc_alternative_products'\n",
    "            }\n",
    "print(','.join([map[c] if c in map else f'{c} not in map' for c in df.columns]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "89fbdbc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entry\n",
      "Reviewed\n",
      "Gene names (primary)\n",
      "Entry name\n",
      "Gene names\n",
      "Organism\n",
      "Length\n",
      "Alternative products\n",
      "Alternative sequence\n",
      "Sequence\n"
     ]
    }
   ],
   "source": [
    "fils = {'Entry': 'accession',\n",
    "            'Absorption': 'absorption',\n",
    "            'Annotation': 'annotation_score',\n",
    "            'Activity regulation': 'cc_activity_regulation',\n",
    "            'Allergenic properties': 'cc_allergen',\n",
    "            'Alternative products': 'cc_alternative_products',\n",
    "            'Biotechnological use': 'cc_biotechnology',\n",
    "            'Catalytic activity': 'cc_catalytic_activity',\n",
    "            'Caution': 'cc_caution',\n",
    "            'Cofactor': 'cc_cofactor',\n",
    "            'Developmental stage': 'cc_developmental_stage',\n",
    "            'Involvement in disease': 'cc_disease',\n",
    "            'Disruption phenotype': 'cc_disruption_phenotype',\n",
    "            'Domain[CC]': 'cc_domain',\n",
    "            'Function [CC]': 'cc_function',\n",
    "            'Induction': 'cc_induction',\n",
    "            'Interacts with': 'cc_interaction',\n",
    "            'Mass spectrometry': 'cc_mass_spectrometry',\n",
    "            'Miscellaneous [CC]': 'cc_miscellaneous',\n",
    "            'Pathway': 'cc_pathway',\n",
    "            'Pharmaceutical use': 'cc_pharmaceutical',\n",
    "            'Polymorphism': 'cc_polymorphism',\n",
    "            'Post-translational modification': 'cc_ptm',\n",
    "            'RNA editing': 'cc_rna_editing',\n",
    "            'Sequence caution': 'cc_sequence_caution',\n",
    "            'Subcellular location[CC]': 'cc_subcellular_location',\n",
    "            'Subunit structure[CC]': 'cc_subunit',\n",
    "            'Tissue specificity': 'cc_tissue_specificity',\n",
    "            'Toxic dose': 'cc_toxic_dose',\n",
    "            'Comment Count': 'comment_count',\n",
    "            'Date of creation': 'date_created',\n",
    "            'Date of last modification': 'date_modified',\n",
    "            'Date of last sequence modification': 'date_sequence_modified',\n",
    "            'EC number': 'ec',\n",
    "            'Erroneous gene model prediction': 'error_gmodel_pred',\n",
    "            'Features': 'feature_count',\n",
    "            'Fragment': 'fragment',\n",
    "            'Active site': 'ft_act_site',\n",
    "            'Binding site': 'ft_binding',\n",
    "            # this field name causes problems 2022-08-22\n",
    "            # 'Calcium binding': 'ft_ca_bind',\n",
    "            'Glycosylation': 'ft_carbohyd',\n",
    "            'Chain': 'ft_chain',\n",
    "            'Coiled coil': 'ft_coiled',\n",
    "            'Compositional bias': 'ft_compbias',\n",
    "            'Sequence conflict': 'ft_conflict',\n",
    "            'Cross-link': 'ft_crosslnk',\n",
    "            'Disulfide bond': 'ft_disulfid',\n",
    "            'DNA binding': 'ft_dna_bind',\n",
    "            'Domain[FT]': 'ft_domain',\n",
    "            'Helix': 'ft_helix',\n",
    "            'Initiator methionine': 'ft_init_met',\n",
    "            'Intramembrane': 'ft_intramem',\n",
    "            'Lipidation': 'ft_lipid',\n",
    "            # this field name causes problems 2022-08-22\n",
    "            # 'Metal binding': 'ft_metal',\n",
    "            'Modified residue': 'ft_mod_res',\n",
    "            'Motif': 'ft_motif',\n",
    "            'Mutagenesis': 'ft_mutagen',\n",
    "            'Non-adjacent residues': 'ft_non_cons',\n",
    "            'Non-standard residue': 'ft_non_std',\n",
    "            'Non-terminal residue': 'ft_non_ter',\n",
    "            # this field name causes problems 2022-08-22\n",
    "            # 'Nucleotide binding': 'ft_np_bind',\n",
    "            'Peptide': 'ft_peptide',\n",
    "            'Propeptide': 'ft_propep',\n",
    "            'Region': 'ft_region',\n",
    "            'Repeat': 'ft_repeat',\n",
    "            'Signal peptide': 'ft_signal',\n",
    "            'Site': 'ft_site',\n",
    "            'Beta strand': 'ft_strand',\n",
    "            'Topological domain': 'ft_topo_dom',\n",
    "            'Transit peptide': 'ft_transit',\n",
    "            'Transmembrane': 'ft_transmem',\n",
    "            'Turn': 'ft_turn',\n",
    "            'Sequence uncertainty': 'ft_unsure',\n",
    "            'Alternative sequence': 'ft_var_seq',\n",
    "            'Natural variant': 'ft_variant',\n",
    "            'Zinc finger': 'ft_zn_fing',\n",
    "            'Gene names': 'gene_names',\n",
    "            'Gene names (ordered locus)': 'gene_oln',\n",
    "            'Gene names (ORF)': 'gene_orf',\n",
    "            'Gene names (primary)': 'gene_primary',\n",
    "            'Gene names (synonym)': 'gene_synonym',\n",
    "            'Gene ontology (GO)': 'go',\n",
    "            'Gene ontology (cellular component)': 'go_c',\n",
    "            'Gene ontology (molecular function)': 'go_f',\n",
    "            'Gene ontology IDs': 'go_id',\n",
    "            'Gene ontology (biological process)': 'go_p',\n",
    "            'Entry name': 'id',\n",
    "            'Keywords': 'keyword',\n",
    "            'Keyword ID': 'keywordid',\n",
    "            'Kinetics': 'kinetics',\n",
    "            'Length': 'length',\n",
    "            'Taxonomic lineage': 'lineage',\n",
    "            'Taxonomic lineage (IDs)': 'lineage_ids',\n",
    "            'PubMed ID': 'lit_pubmed_id',\n",
    "            'Mass': 'mass',\n",
    "            'Gene encoded by': 'organelle',\n",
    "            'Organism ID': 'organism_id',\n",
    "            'Organism': 'organism_name',\n",
    "            'pH dependence': 'ph_dependence',\n",
    "            'Protein existence': 'protein_existence',\n",
    "            'Protein families': 'protein_families',\n",
    "            'Protein names': 'protein_name',\n",
    "            'Redox potential': 'redox_potential',\n",
    "            'Reviewed': 'reviewed',\n",
    "            'Rhea ID': 'rhea',\n",
    "            'Sequence': 'sequence',\n",
    "            'Sequence version': 'sequence_version',\n",
    "            '3D': 'structure_3d',\n",
    "            'Temperature dependence': 'temp_dependence',\n",
    "            'Tools': 'tools',\n",
    "            'UniParc': 'uniparc_id',\n",
    "            'Entry version': 'version',\n",
    "            'Virus hosts': 'virus_hosts',\n",
    "            'ABCD': 'xref_abcd',\n",
    "            'Allergome': 'xref_allergome',\n",
    "            'AlphaFoldDB': 'xref_alphafolddb',\n",
    "            'Antibodypedia': 'xref_antibodypedia',\n",
    "            'ArachnoServer': 'xref_arachnoserver',\n",
    "            'Araport': 'xref_araport',\n",
    "            'Bgee': 'xref_bgee',\n",
    "            'BindingDB': 'xref_bindingdb',\n",
    "            'BioCyc': 'xref_biocyc',\n",
    "            'BioGRID': 'xref_biogrid',\n",
    "            'BioGRID-ORCS': 'xref_biogrid-orcs',\n",
    "            'BioMuta': 'xref_biomuta',\n",
    "            'BMRB': 'xref_bmrb',\n",
    "            'BRENDA': 'xref_brenda',\n",
    "            'CarbonylDB': 'xref_carbonyldb',\n",
    "            'CAZy': 'xref_cazy',\n",
    "            'CCDS': 'xref_ccds',\n",
    "            'CDD': 'xref_cdd',\n",
    "            'CGD': 'xref_cgd',\n",
    "            'ChEMBL': 'xref_chembl',\n",
    "            'ChiTaRS': 'xref_chitars',\n",
    "            'CLAE': 'xref_clae',\n",
    "            'CleanEx': 'xref_cleanex',\n",
    "            'CollecTF': 'xref_collectf',\n",
    "            'ComplexPortal': 'xref_complexportal',\n",
    "            'COMPLUYEAST-2DPAGE': 'xref_compluyeast-2dpage',\n",
    "            'ConoServer': 'xref_conoserver',\n",
    "            'CORUM': 'xref_corum',\n",
    "            'CPTAC': 'xref_cptac',\n",
    "            'CPTC': 'xref_cptc',\n",
    "            'CTD': 'xref_ctd',\n",
    "            'dbSNP': 'xref_dbsnp',\n",
    "            'DEPOD': 'xref_depod',\n",
    "            'dictyBase': 'xref_dictybase',\n",
    "            'DIP': 'xref_dip',\n",
    "            'DisGeNET': 'xref_disgenet',\n",
    "            'DisProt': 'xref_disprot',\n",
    "            'DMDM': 'xref_dmdm',\n",
    "            'DNASU': 'xref_dnasu',\n",
    "            'DOSAC-COBS-2DPAGE': 'xref_dosac-cobs-2dpage',\n",
    "            'DrugBank': 'xref_drugbank',\n",
    "            'DrugCentral': 'xref_drugcentral',\n",
    "            'EchoBASE': 'xref_echobase',\n",
    "            'eggNOG': 'xref_eggnog',\n",
    "            'ELM': 'xref_elm',\n",
    "            'EMBL': 'xref_embl',\n",
    "            'Ensembl': 'xref_ensembl',\n",
    "            'EnsemblBacteria': 'xref_ensemblbacteria',\n",
    "            'EnsemblFungi': 'xref_ensemblfungi',\n",
    "            'EnsemblMetazoa': 'xref_ensemblmetazoa',\n",
    "            'EnsemblPlants': 'xref_ensemblplants',\n",
    "            'EnsemblProtists': 'xref_ensemblprotists',\n",
    "            'EPD': 'xref_epd',\n",
    "            'ESTHER': 'xref_esther',\n",
    "            'euHCVdb': 'xref_euhcvdb',\n",
    "            'EvolutionaryTrace': 'xref_evolutionarytrace',\n",
    "            'ExpressionAtlas': 'xref_expressionatlas',\n",
    "            'FlyBase': 'xref_flybase',\n",
    "            'Gene3D': 'xref_gene3d',\n",
    "            'GeneCards': 'xref_genecards',\n",
    "            'GeneID': 'xref_geneid',\n",
    "            'GeneReviews': 'xref_genereviews',\n",
    "            'GeneTree': 'xref_genetree',\n",
    "            'Genevisible': 'xref_genevisible',\n",
    "            'GeneWiki': 'xref_genewiki',\n",
    "            'GenomeRNAi': 'xref_genomernai',\n",
    "            'GlyConnect': 'xref_glyconnect',\n",
    "            'GlyGen': 'xref_glygen',\n",
    "            'Gramene': 'xref_gramene',\n",
    "            'GuidetoPHARMACOLOGY': 'xref_guidetopharmacology',\n",
    "            'HAMAP': 'xref_hamap',\n",
    "            'HGNC': 'xref_hgnc',\n",
    "            'HOGENOM': 'xref_hogenom',\n",
    "            'HPA': 'xref_hpa',\n",
    "            'IDEAL': 'xref_ideal',\n",
    "            'IMGT_GENE-DB': 'xref_imgt_gene-db',\n",
    "            'InParanoid': 'xref_inparanoid',\n",
    "            'IntAct': 'xref_intact',\n",
    "            'InterPro': 'xref_interpro',\n",
    "            'iPTMnet': 'xref_iptmnet',\n",
    "            'jPOST': 'xref_jpost',\n",
    "            'KEGG': 'xref_kegg',\n",
    "            'KO': 'xref_ko',\n",
    "            'LegioList': 'xref_legiolist',\n",
    "            'Leproma': 'xref_leproma',\n",
    "            'MaizeGDB': 'xref_maizegdb',\n",
    "            'MalaCards': 'xref_malacards',\n",
    "            'MANE-Select': 'xref_mane-select',\n",
    "            'MassIVE': 'xref_massive',\n",
    "            'MaxQB': 'xref_maxqb',\n",
    "            'MEROPS': 'xref_merops',\n",
    "            'MetOSite': 'xref_metosite',\n",
    "            'MGI': 'xref_mgi',\n",
    "            'MIM': 'xref_mim',\n",
    "            'MINT': 'xref_mint',\n",
    "            'MoonDB': 'xref_moondb',\n",
    "            'MoonProt': 'xref_moonprot',\n",
    "            'neXtProt': 'xref_nextprot',\n",
    "            'NIAGADS': 'xref_niagads',\n",
    "            'OGP': 'xref_ogp',\n",
    "            'OMA': 'xref_oma',\n",
    "            'OpenTargets': 'xref_opentargets',\n",
    "            'Orphanet': 'xref_orphanet',\n",
    "            'OrthoDB': 'xref_orthodb',\n",
    "            'PANTHER': 'xref_panther',\n",
    "            'PathwayCommons': 'xref_pathwaycommons',\n",
    "            'PATRIC': 'xref_patric',\n",
    "            'PaxDb': 'xref_paxdb',\n",
    "            'PCDDB': 'xref_pcddb',\n",
    "            'PDB': 'xref_pdb',\n",
    "            'PDBsum': 'xref_pdbsum',\n",
    "            'PeptideAtlas': 'xref_peptideatlas',\n",
    "            'PeroxiBase': 'xref_peroxibase',\n",
    "            'Pfam': 'xref_pfam',\n",
    "            'PharmGKB': 'xref_pharmgkb',\n",
    "            'Pharos': 'xref_pharos',\n",
    "            'PHI-base': 'xref_phi-base',\n",
    "            'PhosphoSitePlus': 'xref_phosphositeplus',\n",
    "            'PhylomeDB': 'xref_phylomedb',\n",
    "            'PIR': 'xref_pir',\n",
    "            'PIRSF': 'xref_pirsf',\n",
    "            'PlantReactome': 'xref_plantreactome',\n",
    "            'PomBase': 'xref_pombase',\n",
    "            'PRIDE': 'xref_pride',\n",
    "            'PRINTS': 'xref_prints',\n",
    "            'PRO': 'xref_pro',\n",
    "            'ProDom': 'xref_prodom',\n",
    "            'ProMEX': 'xref_promex',\n",
    "            'PROSITE': 'xref_prosite',\n",
    "            'Proteomes': 'xref_proteomes',\n",
    "            'ProteomicsDB': 'xref_proteomicsdb',\n",
    "            'PseudoCAP': 'xref_pseudocap',\n",
    "            'Reactome': 'xref_reactome',\n",
    "            'REBASE': 'xref_rebase',\n",
    "            'RefSeq': 'xref_refseq',\n",
    "            'REPRODUCTION-2DPAGE': 'xref_reproduction-2dpage',\n",
    "            'RGD': 'xref_rgd',\n",
    "            'RNAct': 'xref_rnact',\n",
    "            'SABIO-RK': 'xref_sabio-rk',\n",
    "            'SASBDB': 'xref_sasbdb',\n",
    "            'SFLD': 'xref_sfld',\n",
    "            'SGD': 'xref_sgd',\n",
    "            'SignaLink': 'xref_signalink',\n",
    "            'SIGNOR': 'xref_signor',\n",
    "            'SMART': 'xref_smart',\n",
    "            'SMR': 'xref_smr',\n",
    "            'STRING': 'xref_string',\n",
    "            'SUPFAM': 'xref_supfam',\n",
    "            'SWISS-2DPAGE': 'xref_swiss-2dpage',\n",
    "            'SwissLipids': 'xref_swisslipids',\n",
    "            'SwissPalm': 'xref_swisspalm',\n",
    "            'TAIR': 'xref_tair',\n",
    "            'TCDB': 'xref_tcdb',\n",
    "            'TIGRFAMs': 'xref_tigrfams',\n",
    "            'TopDownProteomics': 'xref_topdownproteomics',\n",
    "            'TreeFam': 'xref_treefam',\n",
    "            'TubercuList': 'xref_tuberculist',\n",
    "            'UCD-2DPAGE': 'xref_ucd-2dpage',\n",
    "            'UCSC': 'xref_ucsc',\n",
    "            'UniCarbKB': 'xref_unicarbkb',\n",
    "            'UniLectin': 'xref_unilectin',\n",
    "            'UniPathway': 'xref_unipathway',\n",
    "            'VectorBase': 'xref_vectorbase',\n",
    "            'VEuPathDB': 'xref_veupathdb',\n",
    "            'VGNC': 'xref_vgnc',\n",
    "            'WBParaSite': 'xref_wbparasite',\n",
    "            'WBParaSiteTranscriptProtein': 'xref_wbparasitetranscriptprotein',\n",
    "            'World-2DPAGE': 'xref_world-2dpage',\n",
    "            'WormBase': 'xref_wormbase',\n",
    "            'Xenbase': 'xref_xenbase',\n",
    "            'ZFIN': 'xref_zfin'\n",
    "            }\n",
    "fils = {v:k for k,v in fils.items()}\n",
    "for f in [\n",
    "    \"accession\",\n",
    "    \"reviewed\",\n",
    "    \"gene_primary\",\n",
    "    \"id\",\n",
    "    \"gene_names\",\n",
    "    \"organism_name\",\n",
    "    \"length\",\n",
    "    \"cc_alternative_products\",\n",
    "    \"ft_var_seq\",\n",
    "    \"sequence\"\n",
    "]:\n",
    "    print(fils[f])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2d4ad08d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data['results'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "141e8278",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching: https://rest.uniprot.org/taxonomy\n",
      "Error: 404 - <html>\n",
      "<head><title>404 Not Found</title></head>\n",
      "<body>\n",
      "<center><h1>404 Not Found</h1></center>\n",
      "<hr><center>nginx/1.25.3</center>\n",
      "</body>\n",
      "</html>\n",
      "\n",
      "Fetched 0 organisms.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "# Base URL for UniProt API\n",
    "UNIPROT_API_URL = \"https://rest.uniprot.org/taxonomy\"\n",
    "\n",
    "# Parameters for the API request\n",
    "params = {\n",
    "    \"fields\": \"name,common_name,synonym,scientific_name,rank\",\n",
    "    \"format\": \"json\",\n",
    "    \"size\": 500,\n",
    "}\n",
    "\n",
    "def fetch_organisms():\n",
    "    \"\"\"Fetch all organisms from UniProt API and save them to a JSON file.\"\"\"\n",
    "    all_organisms = []\n",
    "    url = UNIPROT_API_URL\n",
    "    \n",
    "    while url:\n",
    "        print(f\"Fetching: {url}\")\n",
    "        response = requests.get(url, params=params if url == UNIPROT_API_URL else None)\n",
    "        \n",
    "        if response.status_code != 200:\n",
    "            print(f\"Error: {response.status_code} - {response.text}\")\n",
    "            break\n",
    "        \n",
    "        data = response.json()\n",
    "        all_organisms.extend(data.get(\"results\", []))\n",
    "        url = data.get(\"next\")  # URL for the next page\n",
    "\n",
    "    print(f\"Fetched {len(all_organisms)} organisms.\")\n",
    "    return all_organisms\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    fetch_organisms()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f411561",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "168be1d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14456"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['Organism'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4d13978e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Entry', 'Reviewed', 'Gene Names (primary)', 'Entry Name', 'Gene Names',\n",
       "       'Organism', 'Length', 'Alternative products (isoforms)',\n",
       "       'Alternative sequence', 'Sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9760aa11-8796-4874-adb1-855b0151cb15",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "db_file = os.path.join('data','db','proteogyver.db')\n",
    "db_conn = db_functions.create_connection(db_file)\n",
    "new_current = db_functions.get_full_table_as_pd(db_conn, 'common_proteins')\n",
    "db_conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7263150f",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_modify = []\n",
    "for _,row in current[current['protein_type'].str.contains('ipynb')].iterrows():\n",
    "    to_modify.append([row['uniprot_id'], row['protein_type'].replace('.ipynb_checkpoints, ', '')])\n",
    "db_conn = db_functions.create_connection(db_file)\n",
    "for mod in to_modify:\n",
    "    db_functions.modify_record(db_conn, 'common_proteins', 'uniprot_id', mod[0], ['protein_type'], [mod[1]])\n",
    "db_conn.commit()\n",
    "db_conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ca826053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_type\n",
       "Proteasome                                           474934\n",
       "Keratin                                               75348\n",
       "Elongation factor                                      5290\n",
       "Ribosome                                               3942\n",
       "Carboxylase                                            2816\n",
       "Chaperone                                              2351\n",
       "Histone                                                2017\n",
       "Heat shock protein                                     1373\n",
       "Actin                                                  1207\n",
       "Chaperone, Heat shock protein                          1135\n",
       "Tubulin                                                 671\n",
       "Heat shock protein, Proteasome                          117\n",
       "Chaperone, Histone                                       89\n",
       "Keratin, Proteasome                                      71\n",
       "Elongation factor, Ribosome                              40\n",
       "Chaperone, Proteasome                                    40\n",
       "Chaperone, Tubulin                                       38\n",
       "Histone, Proteasome                                      31\n",
       "Elongation factor, Proteasome                            24\n",
       "Actin, Proteasome                                        23\n",
       "Proteasome, Ribosome                                     12\n",
       "Chaperone, Heat shock protein, Ribosome                  11\n",
       "Carboxylase, Chaperone                                    7\n",
       "Proteasome, Tubulin                                       6\n",
       "Carboxylase, Histone                                      5\n",
       "Carboxylase, Proteasome                                   3\n",
       "Chaperone, Histone, Proteasome                            3\n",
       "Chaperone, Proteasome, Tubulin                            3\n",
       "Chaperone, Heat shock protein, Proteasome                 3\n",
       "Heat shock protein, Ribosome                              2\n",
       "Actin, Elongation factor                                  2\n",
       "Histone, Proteasome, Tubulin                              2\n",
       "Actin, Heat shock protein                                 1\n",
       "Heat shock protein, Keratin                               1\n",
       "Heat shock protein, Histone                               1\n",
       "Carboxylase, Elongation factor                            1\n",
       "Chaperone, Elongation factor, Histone, Proteasome         1\n",
       "Actin, Keratin                                            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_current['protein_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "107cb0c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "571622"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "current['protein_type'].value_counts().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e53def19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>entry_name</th>\n",
       "      <th>all_gene_names</th>\n",
       "      <th>organism</th>\n",
       "      <th>protein_type</th>\n",
       "      <th>update_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P04792</td>\n",
       "      <td>HSPB1</td>\n",
       "      <td>HSPB1_HUMAN</td>\n",
       "      <td>HSPB1 HSP27 HSP28</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Heat shock protein, Proteasome</td>\n",
       "      <td>2024-11-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P63284</td>\n",
       "      <td>clpB</td>\n",
       "      <td>CLPB_ECOLI</td>\n",
       "      <td>clpB htpM b2592 JW2573</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>Chaperone, Heat shock protein</td>\n",
       "      <td>2024-11-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P10591</td>\n",
       "      <td>SSA1</td>\n",
       "      <td>HSP71_YEAST</td>\n",
       "      <td>SSA1 YAL005C</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 /...</td>\n",
       "      <td>Heat shock protein, Proteasome</td>\n",
       "      <td>2024-11-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q00613</td>\n",
       "      <td>HSF1</td>\n",
       "      <td>HSF1_HUMAN</td>\n",
       "      <td>HSF1 HSTF1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Heat shock protein, Proteasome</td>\n",
       "      <td>2024-11-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O14558</td>\n",
       "      <td>HSPB6</td>\n",
       "      <td>HSPB6_HUMAN</td>\n",
       "      <td>HSPB6</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Heat shock protein</td>\n",
       "      <td>2024-11-21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id gene_name   entry_name          all_gene_names  \\\n",
       "0     P04792     HSPB1  HSPB1_HUMAN       HSPB1 HSP27 HSP28   \n",
       "1     P63284      clpB   CLPB_ECOLI  clpB htpM b2592 JW2573   \n",
       "2     P10591      SSA1  HSP71_YEAST            SSA1 YAL005C   \n",
       "3     Q00613      HSF1   HSF1_HUMAN              HSF1 HSTF1   \n",
       "4     O14558     HSPB6  HSPB6_HUMAN                   HSPB6   \n",
       "\n",
       "                                            organism  \\\n",
       "0                               Homo sapiens (Human)   \n",
       "1                      Escherichia coli (strain K12)   \n",
       "2  Saccharomyces cerevisiae (strain ATCC 204508 /...   \n",
       "3                               Homo sapiens (Human)   \n",
       "4                               Homo sapiens (Human)   \n",
       "\n",
       "                     protein_type update_time  \n",
       "0  Heat shock protein, Proteasome  2024-11-21  \n",
       "1   Chaperone, Heat shock protein  2024-11-21  \n",
       "2  Heat shock protein, Proteasome  2024-11-21  \n",
       "3  Heat shock protein, Proteasome  2024-11-21  \n",
       "4              Heat shock protein  2024-11-21  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "current.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cea59df-9470-4d20-88cf-3f8a272f0565",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>entry_name</th>\n",
       "      <th>all_gene_names</th>\n",
       "      <th>organism</th>\n",
       "      <th>protein_type</th>\n",
       "      <th>update_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1372</th>\n",
       "      <td>P27449</td>\n",
       "      <td>ATP6V0C</td>\n",
       "      <td>VATL_HUMAN</td>\n",
       "      <td>ATP6V0C ATP6C ATP6L ATPL</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Proteasome components</td>\n",
       "      <td>2024-02-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3407</th>\n",
       "      <td>Q9NQ11</td>\n",
       "      <td>ATP13A2</td>\n",
       "      <td>AT132_HUMAN</td>\n",
       "      <td>ATP13A2 PARK9</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Proteasome components</td>\n",
       "      <td>2024-02-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3966</th>\n",
       "      <td>P23956</td>\n",
       "      <td>ATP6V0C</td>\n",
       "      <td>VATL_BOVIN</td>\n",
       "      <td>ATP6V0C ATP6C ATP6L</td>\n",
       "      <td>Bos taurus (Bovine)</td>\n",
       "      <td>Proteasome components</td>\n",
       "      <td>2024-02-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5297</th>\n",
       "      <td>O18882</td>\n",
       "      <td>ATP6V0C</td>\n",
       "      <td>VATL_SHEEP</td>\n",
       "      <td>ATP6V0C ATP6C ATP6L</td>\n",
       "      <td>Ovis aries (Sheep)</td>\n",
       "      <td>Proteasome components</td>\n",
       "      <td>2024-02-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7747</th>\n",
       "      <td>Q9LU63</td>\n",
       "      <td>ATP1</td>\n",
       "      <td>ATP1_ARATH</td>\n",
       "      <td>ATP1 PDL1 SDIRIP1 At5g51110 MWD22.5</td>\n",
       "      <td>Arabidopsis thaliana (Mouse-ear cress)</td>\n",
       "      <td>Carboxylases</td>\n",
       "      <td>2024-02-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     uniprot_id gene_name   entry_name                       all_gene_names  \\\n",
       "1372     P27449   ATP6V0C   VATL_HUMAN             ATP6V0C ATP6C ATP6L ATPL   \n",
       "3407     Q9NQ11   ATP13A2  AT132_HUMAN                        ATP13A2 PARK9   \n",
       "3966     P23956   ATP6V0C   VATL_BOVIN                  ATP6V0C ATP6C ATP6L   \n",
       "5297     O18882   ATP6V0C   VATL_SHEEP                  ATP6V0C ATP6C ATP6L   \n",
       "7747     Q9LU63      ATP1   ATP1_ARATH  ATP1 PDL1 SDIRIP1 At5g51110 MWD22.5   \n",
       "\n",
       "                                    organism           protein_type  \\\n",
       "1372                    Homo sapiens (Human)  Proteasome components   \n",
       "3407                    Homo sapiens (Human)  Proteasome components   \n",
       "3966                     Bos taurus (Bovine)  Proteasome components   \n",
       "5297                      Ovis aries (Sheep)  Proteasome components   \n",
       "7747  Arabidopsis thaliana (Mouse-ear cress)           Carboxylases   \n",
       "\n",
       "     update_time  \n",
       "1372  2024-02-16  \n",
       "3407  2024-02-16  \n",
       "3966  2024-02-16  \n",
       "5297  2024-02-16  \n",
       "7747  2024-02-16  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "current[current['gene_name'].fillna('').str.contains('ATP')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87614137-c8d0-4273-8cda-8e3c4886cdd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "comdir = os.path.join('data_assets','db build files','Potential contaminant protein groups')\n",
    "comdir = '/home/kmsaloka/Downloads/Database additions/Potential contaminant protein groups'\n",
    "db_file = os.path.join('data','proteogyver.db')\n",
    "\n",
    "com_cols = [\n",
    "    'uniprot_id TEXT PRIMARY KEY',\n",
    "    'gene_name TEXT',\n",
    "    'entry_name TEXT',\n",
    "    'all_gene_names TEXT',\n",
    "    'organism TEXT',\n",
    "    'protein_type TEXT NOT NULL',\n",
    "    'update_time TEXT NOT NULL '\n",
    "]\n",
    "db_conn = db_functions.create_connection(db_file)\n",
    "current = db_functions.get_full_table_as_pd(db_conn, 'common_proteins')\n",
    "have = set(current['uniprot_id'].values)\n",
    "db_conn.close()\n",
    "common_proteins = {}\n",
    "for root, dirs, files in os.walk(comdir):\n",
    "    for f in files:\n",
    "        comdf = pd.read_csv(os.path.join(root,f),sep='\\t')\n",
    "        name = root.rsplit(os.sep,maxsplit=1)[-1]\n",
    "        for _,row in comdf.iterrows():\n",
    "            if row['Entry'] not in common_proteins:\n",
    "                common_proteins[row['Entry']] = [\n",
    "                    row['Entry'],\n",
    "                    row['Gene Names (primary)'],\n",
    "                    row['Entry Name'],\n",
    "                    row['Gene Names'],\n",
    "                    row['Organism'],\n",
    "                    [name],\n",
    "                    datetime.today().strftime('%Y-%m-%d')\n",
    "                ]\n",
    "            else:\n",
    "                common_proteins[row['Entry']][5].append(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13283faf-9f0e-475d-89a7-58ada01cafea",
   "metadata": {},
   "outputs": [],
   "source": [
    "modifications = []\n",
    "additions = []\n",
    "for common_protein, data in common_proteins.items():\n",
    "    if common_protein in have:\n",
    "        cur_ver = current[current['uniprot_id']==common_protein]\n",
    "        mods = cur_ver['protein_type'].split(', ')\n",
    "        mods.extend(data[5])\n",
    "        modifications.append([common_protein, 'protein_type', ', '.join(sorted(list(set(mods))))])\n",
    "    else:\n",
    "        data[5] = ', '.join(data[5])\n",
    "        additions.append([com_cols, data])\n",
    "db_conn = db_functions.create_connection(db_file)\n",
    "for addcols, add_data in additions:\n",
    "    db_functions.add_record(db_conn, 'common_proteins', addcols, add_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1accc3a3-0491-4429-862c-ec24827d7e42",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>entry_name</th>\n",
       "      <th>all_gene_names</th>\n",
       "      <th>organism</th>\n",
       "      <th>protein</th>\n",
       "      <th>update_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0G2JZ79</td>\n",
       "      <td>Sirt1</td>\n",
       "      <td>SIR1_RAT</td>\n",
       "      <td>Sirt1</td>\n",
       "      <td>Rattus norvegicus (Rat)</td>\n",
       "      <td>Keratin</td>\n",
       "      <td>2024-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0H2URK1</td>\n",
       "      <td>psrP</td>\n",
       "      <td>PSRP_STRPN</td>\n",
       "      <td>psrP SP_1772</td>\n",
       "      <td>Streptococcus pneumoniae serotype 4 (strain AT...</td>\n",
       "      <td>Keratin</td>\n",
       "      <td>2024-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D3ZER2</td>\n",
       "      <td>Bfsp2</td>\n",
       "      <td>BFSP2_RAT</td>\n",
       "      <td>Bfsp2</td>\n",
       "      <td>Rattus norvegicus (Rat)</td>\n",
       "      <td>Keratin</td>\n",
       "      <td>2024-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F7EQ49</td>\n",
       "      <td>GPER1</td>\n",
       "      <td>GPER1_MACMU</td>\n",
       "      <td>GPER1 GPER GPR30</td>\n",
       "      <td>Macaca mulatta (Rhesus macaque)</td>\n",
       "      <td>Keratin</td>\n",
       "      <td>2024-02-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O08878</td>\n",
       "      <td>Gper1</td>\n",
       "      <td>GPER1_RAT</td>\n",
       "      <td>Gper1 Cmkrl2 Gper Gpr30 Gpr41</td>\n",
       "      <td>Rattus norvegicus (Rat)</td>\n",
       "      <td>Keratin</td>\n",
       "      <td>2024-02-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id gene_name   entry_name                 all_gene_names  \\\n",
       "0  A0A0G2JZ79     Sirt1     SIR1_RAT                          Sirt1   \n",
       "1  A0A0H2URK1      psrP   PSRP_STRPN                   psrP SP_1772   \n",
       "2      D3ZER2     Bfsp2    BFSP2_RAT                          Bfsp2   \n",
       "3      F7EQ49     GPER1  GPER1_MACMU               GPER1 GPER GPR30   \n",
       "4      O08878     Gper1    GPER1_RAT  Gper1 Cmkrl2 Gper Gpr30 Gpr41   \n",
       "\n",
       "                                            organism  protein update_time  \n",
       "0                            Rattus norvegicus (Rat)  Keratin  2024-02-05  \n",
       "1  Streptococcus pneumoniae serotype 4 (strain AT...  Keratin  2024-02-05  \n",
       "2                            Rattus norvegicus (Rat)  Keratin  2024-02-05  \n",
       "3                    Macaca mulatta (Rhesus macaque)  Keratin  2024-02-05  \n",
       "4                            Rattus norvegicus (Rat)  Keratin  2024-02-05  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "current.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "57c1fd6d-53c3-46bb-8233-65df886139ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def rename_column(db_conn, tablename, old_col, new_col):\n",
    "    sql_str = f'ALTER TABLE {tablename} RENAME COLUMN {old_col} TO {new_col};'\n",
    "    try:\n",
    "        db_conn.cursor().execute(sql_str)\n",
    "    except sqlite3.Error as error:\n",
    "        print(f'Failed to rename column in sqlite table. Error: {error}', sql_str)\n",
    "        raise\n",
    "db_file = os.path.join('data','proteogyvertesting.db')\n",
    "con = db_functions.create_connection(db_file)\n",
    "rename_column(con, 'common_proteins','protein','protein_type')\n",
    "con.commit()\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f27ad91f-ddde-44d7-9996-666a29a28557",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('control_sets',), ('crapome_sets',), ('proteins',), ('control_vl_gfp_mac3_10min_ap_overall',), ('control_vl_gfp_mac3_10min_ap',), ('control_vl_gfp_mac3_10min_bioid_overall',), ('control_vl_gfp_mac3_10min_bioid',), ('control_vl_gfp_mac2_18h_ap_overall',), ('control_vl_gfp_mac2_18h_ap',), ('control_vl_gfp_mac2_18h_bioid_overall',), ('control_vl_gfp_mac2_18h_bioid',), ('control_vl_gfp_mac_24h_ap_overall',), ('control_vl_gfp_mac_24h_ap',), ('control_vl_gfp_mac_24h_ap_nls_overall',), ('control_vl_gfp_mac_24h_ap_nls',), ('control_vl_gfp_mac_24h_bioid_overall',), ('control_vl_gfp_mac_24h_bioid',), ('control_vl_gfp_mac_24h_bioid_nls_overall',), ('control_vl_gfp_mac_24h_bioid_nls',), ('crapome_vl_gfp_mac3_10min_ap',), ('crapome_vl_gfp_mac3_10min_bioid',), ('crapome_vl_gfp_mac2_18h_ap',), ('crapome_vl_gfp_mac2_18h_bioid',), ('crapome_vl_gfp_mac_24h_ap',), ('crapome_vl_gfp_mac_24h_ap_nls',), ('crapome_vl_gfp_mac_24h_bioid',), ('crapome_vl_gfp_mac_24h_bioid_nls',), ('crapome_nesvilab',), ('contaminants',), ('ms_runs',), ('known_interactions',), ('msmicroscopy',), ('common_proteins',)]\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "con = sqlite3.connect(db_file)\n",
    "cursor = con.cursor()\n",
    "cursor.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "print(cursor.fetchall())\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5354dccb-4602-444a-8b8f-d537e1975e3b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
